### ISSUE TITLE
CreateTokenEffect needs to be refactored to allow creating multiple tokens in on CreateTokenEvent

### ISSUE THREAD

**Issue by Grath** [magefree/mage#10811]
Issue:
When multiple tokens of a different type are created such as by [[Trostani's Summoner]], they currently create multiple `CreateTokenEvent`s which cause problems with replacement effects such as [[Queen Allenal of Ruadach]] and [[Peregrin Took]] which add more additional tokens than they should.

Proposed solution:
- Create a `TokenBatch` or similar, which can contain a collection of `Token`s and `CreateTokenCopyTargetEffect`s (or something else that can specify "Create a token copy that's different" for [[Soul Separator]] support) which are then `putOntoTheBattlefield` in a single `CreateTokenEvent` which already supports creating a collection of tokens.
- Refactor `CreateTokenEffect` to take a `TokenBatch` instead of `Token`, retaining the `Token` constructors with a "encase the `Token` in a `TokenBatch`, then use the `TokenBatch` constructors" to reduce duplicated code rather than making a separate `CreateTokenBatchEffect` but not requiring every `CreateTokenEffect` to be rewritten.
- Rewrite (or initially write) the following cards (and possibly more) to use `TokenBatch`es to properly create all their multiple tokens in a single `CreateTokenEvent`:
Bestial Menace, Farmer Cotton, Fae Offering, Battle for Bretagard, Forbidden Friendship, Old Rutstein, Somberwald Beastmaster, Song of Earendil, probably Stangg Echo Warrior, Specimen Collector, Soul Separator, The Eleventh Hour, Triplicate Titan, Trostani's Summoner, Wurmcoil Engine.

**Comment by Susucre** [comment]
[[Woodland Champion]] and [[Kambal, Profiteering Mayor]] have triggers that will also have bugs until the token creation are batched properly.
Champion does trigger 3 times on [[Bestial Menace]] resolution (each trigger seeing 1 token), although it should only be 1 (that sees 3 tokens). And Kambal on opponent's Bestial Menace resolution should create all 3 kind of tokens, currently is only creating the first one (1/1 Snake).

**Comment by JayDi85** [comment]
Original problem in multiple effects usage (game engine processing each ability’s effect as independent step and call full game cycle after each effect).

So alternative fix: create CreateMultipleTokensEffect and call CreateTokenEffect.apply for each token type inside.

P.S. Alternative fix can be impossible due ApplyEffects usage on move to battlefield code. Need research.

**Comment by JayDi85** [comment]
Be careful, Soul Separator doesn't use one token effect — it’s two different:

The Spirit token will be on the battlefield before the Zombie token enters the battlefield.
(2016-07-13)

**Comment by Grath** [comment]
Yeah, I've been advised on the correct behavior (each 'Create' is a separate token creation event to which replacement effects can be applied). For now I'm just handling the 'make multiple pre-defined tokens' and then I'll make a separate pull request for 'make multiple copies' which will not touch Soul Separator.
